core/src/main/java/org/apache/iceberg/io/CachingFileIO.java

```diff
@@ -0,0 +1,355 @@
+    } else {
+      return null;
+    }
+  }
+
+  public static synchronized ContentCache getOrCreate(Map<String, String> properties) {
+    if (sharedCache == null) {
+      sharedCache = createCache(properties);
+    }
+    return sharedCache;
+  }
+
+  public static boolean ioCacheEnabled(Map<String, String> properties) {
+    boolean enabled = PropertyUtil.propertyAsBoolean(properties,
+        CatalogProperties.IO_CACHE_ENABLED,
+        CatalogProperties.IO_CACHE_ENABLED_DEFAULT);
+    long durationMs = PropertyUtil.propertyAsLong(properties,
+        CatalogProperties.IO_CACHE_EXPIRATION_INTERVAL_MS,
+        CatalogProperties.IO_CACHE_EXPIRATION_INTERVAL_MS_DEFAULT);
+    long totalBytes = PropertyUtil.propertyAsLong(properties,
+        CatalogProperties.IO_CACHE_MAX_TOTAL_BYTES,
+        CatalogProperties.IO_CACHE_MAX_TOTAL_BYTES_DEFAULT);
+
+    return enabled && durationMs >= 0 && totalBytes > 0;
+  }
+
+  private ContentCache fileContentCache;
+  private Map<String, String> properties;
+  private FileIO wrappedIO;
+
+  /**
+   * CachingFileIO is not meant to be dynamically loaded.
+   */
+  private CachingFileIO() {
+  }
+
+  public CachingFileIO(FileIO io) {
+    assignIO(io);
+  }
+
+  public void assignIO(FileIO io) {
+    this.wrappedIO = io;
+  }
+
+  /**
+   * Initialize File IO from catalog properties.
+   * <p>
+   * It assumes that the inner FileIO is already initialized and therefore does not reinitialize it.
+   * If this is a reinitialization, it is the caller responsibility to call {@link #close()} and
+   * {@link #assignIO(FileIO)} prior to reinitialization to close the previously assigned FileIO and reassign a new one.
+   * Otherwise, previously assigned FileIO will be kept in use.
+   * <p>
+   * Reinitialization will discard previous content cache unless shared cache is being used
+   * ({@link CatalogProperties#IO_CACHE_SHARED} stays true).
+   * @param newProperties catalog properties
+   */
+  @Override
+  public void initialize(Map<String, String> newProperties) {
+    this.properties = newProperties;
+    initCache();
+  }
+
+  private void initCache() {
+    boolean isSharedCache = PropertyUtil.propertyAsBoolean(properties,
+        CatalogProperties.IO_CACHE_SHARED,
+        CatalogProperties.IO_CACHE_SHARED_DEFAULT);
+    if (isSharedCache) {
+      this.fileContentCache = getOrCreate(properties);
+      LOG.info("CachingFileIO created with shared cache. {} ", this.fileContentCache);
+    } else {
+      this.fileContentCache = createCache(properties);
+      LOG.info("CachingFileIO created. {} ", this.fileContentCache);
+    }
+  }
+
+  @Override
+  public void close() {
+    if (wrappedIO != null) {
+      wrappedIO.close();
+      wrappedIO = null;
+    }
+  }
+
+  @Override
+  public InputFile newInputFile(String path) {
+    InputFile inFile = wrappedIO.newInputFile(path);
+    if (fileContentCache != null && path.contains("/metadata/")) {
```

Comment:
Should this be a config setting since there is a TODO item about it? A regex could be really useful.

Reply:
Hi @danielcweeks , thank you for your feedback!

I can't seem to find an easy way to pass catalog properties (where we define caching related knobs) to ManifestFiles.java. How about we let the CachingFileIO wrapping just like now, but create new method, say `newMetadataInputFile()`, that specially used by `ManifestReader.read()`?

Maybe modification like this in ManifestFiles.java:85
```
InputFile file = (io instanceof CachingFileIO)
    ? ((CachingFileIO) io).newMetadataInputFile(manifest.path())
    : io.newInputFile(manifest.path());
```

For the `newInputFile()` implementation by CachingFileIO, we can just pass-through it to the inner FileIO. Thus, it ensure that caching only applies to metadata files.

Reply:
After looking around, I ended up go the regex route.

There seems to be many call sites where Iceberg is reading metadata files, not just in ManifestFiles.java. Among them are:
```
$ git grep -e "newInputFile.*manifest" -e "newInputFile.*meta"
core/src/main/java/org/apache/iceberg/AllManifestsTable.java:              .withInputFile(io.newInputFile(snap.manifestListLocation()))
core/src/main/java/org/apache/iceberg/AllManifestsTable.java:              io.newInputFile(tableOps().current().metadataFileLocation()),
core/src/main/java/org/apache/iceberg/AllManifestsTable.java:          .read(io.newInputFile(manifestListTask.file().path().toString()))
core/src/main/java/org/apache/iceberg/BaseRewriteManifests.java:    InputFile toCopy = ops.io().newInputFile(manifest.path());
core/src/main/java/org/apache/iceberg/BaseSnapshot.java:      this.allManifests = ManifestLists.read(io.newInputFile(manifestListLocation));
core/src/main/java/org/apache/iceberg/FastAppend.java:    InputFile toCopy = ops.io().newInputFile(manifest.path());
core/src/main/java/org/apache/iceberg/HistoryTable.java:        ops.io().newInputFile(ops.current().metadataFileLocation()),
core/src/main/java/org/apache/iceberg/ManifestFiles.java:    InputFile file = io.newInputFile(manifest.path());
core/src/main/java/org/apache/iceberg/ManifestFiles.java:    InputFile file = io.newInputFile(manifest.path());
core/src/main/java/org/apache/iceberg/ManifestsTable.java:        ops.io().newInputFile(location != null ? location : ops.current().metadataFileLocation()),
core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:    InputFile toCopy = ops.io().newInputFile(manifest.path());
core/src/main/java/org/apache/iceberg/PartitionsTable.java:          io().newInputFile(ops.current().metadataFileLocation()),
core/src/main/java/org/apache/iceberg/PartitionsTable.java:          io().newInputFile(ops.current().metadataFileLocation()),
core/src/main/java/org/apache/iceberg/RemoveSnapshots.java:      return Avro.read(ops.io().newInputFile(snapshot.manifestListLocation()))
core/src/main/java/org/apache/iceberg/SnapshotsTable.java:        ops.io().newInputFile(ops.current().metadataFileLocation()),
hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:    InputFile metadataFile = fileIO.newInputFile(metadataFileLocation);
```